controllers/generic_service_handler.go

```diff
@@ -0,0 +1,149 @@
+package controllers
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/aiven/aiven-go-client"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/aiven/aiven-operator/api/v1alpha1"
+)
+
+func newGenericServiceHandler(fabric serviceAdapterFabric) Handlers {
+	return &genericServiceHandler{fabric: fabric}
+}
+
+// genericServiceHandler provides common CRUD management for all service types using serviceAdapter,
+// which turns specific service (mysql, redis) into a generic.
+type genericServiceHandler struct {
+	fabric serviceAdapterFabric
+}
+
+func (h *genericServiceHandler) createOrUpdate(a *aiven.Client, object client.Object, refs []client.Object) error {
+	o, err := h.fabric(object)
+	if err != nil {
+		return err
+	}
+
+	spec := o.getServiceCommonSpec()
+
+	// Project id reference
+	// Could be right in spec or referenced (has ref)
+	projectVPCID := spec.ProjectVPCID
+	if projectVPCID == "" {
+		if p := v1alpha1.FindProjectVPC(refs); p != nil {
+			projectVPCID = p.Status.ID
+		}
+	}
+
+	_, err = a.Services.Get(spec.Project, o.getServiceName())
+	exists := err == nil
+	if !exists && !aiven.IsNotFound(err) {
+		return fmt.Errorf("failed to fetch service: %w", err)
+	}
+
+	// Creates if not exists or updates existing service
+	var reason string
+	if !exists {
+		reason = "Created"
+		req := o.newCreateRequest()
+		req.ProjectVPCID = toOptionalStringPointer(projectVPCID)
```

Comment:
to deduplicate this part a bit, create a `projectVPCIDPtr` variable before `if !exists` and use it in the block instead:

```suggestion
		req.ProjectVPCID = projectVPCIDPtr
```

Reply:
I believe that doesn't make any difference. It is still a duplicated assignment.

Reply:
this is easier to break as it adds an extra branch—if someone wants to change the way this functions, they can easily forget about the other line, so please change it

Reply:
See my above answer.